src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1285,6 +1302,85 @@ void IMergeTreeDataPart::loadRowsCount()
     }
 }
 
+void IMergeTreeDataPart::readExistingRowsCount() const
+{
+    if (!supportLightweightDeleteMutate() || !hasLightweightDelete() || !storage.getSettings()->exclude_deleted_rows_for_part_size_in_merge
+        || existing_rows_count < rows_count || !getMarksCount())
+        return;
+
+    std::lock_guard lock(existing_rows_count_mutex);
+
+    /// Already read by another thread
+    if (existing_rows_count < rows_count)
+        return;
+
+    NamesAndTypesList cols;
+    cols.push_back(LightweightDeleteDescription::FILTER_COLUMN);
+
+    StorageMetadataPtr metadata_ptr = storage.getInMemoryMetadataPtr();
+    StorageSnapshotPtr storage_snapshot_ptr = std::make_shared<StorageSnapshot>(storage, metadata_ptr);
+
+    MergeTreeReaderPtr reader = getReader(
+        cols,
+        storage_snapshot_ptr,
+        MarkRanges{MarkRange(0, getMarksCount())},
+        nullptr,
+        storage.getContext()->getMarkCache().get(),
+        std::make_shared<AlterConversions>(),
+        MergeTreeReaderSettings{},
+        ValueSizeMap{},
+        ReadBufferFromFileBase::ProfileCallback{});
+
+    if (!reader)
+    {
+        LOG_WARNING(storage.log, "Create reader failed while reading existing rows count");
+        existing_rows_count = rows_count;
+        return;
+    }
+
+    size_t current_mark = 0;
+    const size_t total_mark = getMarksCount();
+
+    bool continue_reading = false;
+    size_t current_row = 0;
+    size_t existing_count = 0;
+
+    while (current_row < rows_count)
+    {
+        size_t rows_to_read = index_granularity.getMarkRows(current_mark);
+        continue_reading = (current_mark != 0);
+
+        Columns result;
+        result.resize(1);
+
+        size_t rows_read = reader->readRows(current_mark, total_mark, continue_reading, rows_to_read, result);
+        if (!rows_read)
+        {
+            LOG_WARNING(storage.log, "Part {} has lightweight delete, but _row_exists column not found", name);
+            existing_rows_count = rows_count;
+            return;
+        }
+
+        current_row += rows_read;
+        current_mark += (rows_to_read == rows_read);
```

Comment:
this doesn't look correct - `rows_read` apparently can come out less than `rows_to_read` - in this case `current_mark` will not be advanced - while it doesn't matter for other than first read (if `continue_reading` is true then `current_mark` doesn't matter), for the first read it does matter and if on the first read we get less than `rows_to_read` then `current_mark` doesn't advance then on the next iteration `continue_reading` will remain false and reading will restart from mark 0

Reply:
I implemented this part with reference to [MergeTreeSequentialSource::generate()](https://github.com/ClickHouse/ClickHouse/blob/f503ea27e9a98d90c385ae2706ebfb5ca9a6315c/src/Storages/MergeTree/MergeTreeSequentialSource.cpp#L170C23-L185)
It looks like `reader->readRows()` will read all rows in this mark range (or up to specified `max_rows_to_read`)?